File: scripts/fetch-job-export.py
#!/usr/bin/env python3
"""
Fetch job export from CampOS API and generate jobs-export.json

This script fetches data from the CampOS API:
- Job export: Contains job data with organization references
- Organization export: Contains organization hierarchy (parent_path)

And generates:
- jobs-export.json (for use by the jobbank Vue app)

Configuration:
- API credentials are stored in config.json (not committed to git)
- See config.example.json for the required structure

Usage:
    python3 scripts/fetch-job-export.py
"""

import json
import os
import re
import sys
from datetime import datetime
from pathlib import Path
from urllib.request import urlopen, Request
from urllib.error import URLError, HTTPError


class JobExportFetcher:
    """Fetch job export data from CampOS API and generate JSON"""

    # Organization name overrides - matches config/org-overrides.js
    ORG_NAME_OVERRIDES = {
        'Havet': 'Havet (underlejr)',
        'Skoven': 'Skoven (underlejr)',
        'Byen': 'Byen (underlejr)',
        'Landet': 'Landet (underlejr)',
    }

    def __init__(self, config_file, output_file):
        self.config_file = config_file
        self.output_file = output_file
        self.org_lookup = {}
        self.config = None

    def load_config(self):
        """Load API configuration from config.json"""
        try:
            with open(self.config_file, 'r', encoding='utf-8') as f:
                self.config = json.load(f)

            # Validate required fields
            required_fields = ['job_export_url', 'org_export_url']
            missing_fields = [f for f in required_fields if f not in self.config]
            if missing_fields:
                print(f"Error: Missing required fields in {self.config_file}: {', '.join(missing_fields)}")
                print("Please see config.example.json for the required structure")
                sys.exit(1)

            print(f"Loaded configuration from {self.config_file}")
            return True

        except FileNotFoundError:
            print(f"Error: Configuration file not found: {self.config_file}")
            print("Please create config.json based on config.example.json")
            sys.exit(1)
        except json.JSONDecodeError as e:
            print(f"Error: Invalid JSON in {self.config_file}: {e}")
            sys.exit(1)

    def fetch_json(self, url):
        """Fetch JSON data from a URL"""
        try:
            req = Request(url)
            req.add_header('User-Agent', 'JobBank/1.0')

            with urlopen(req, timeout=30) as response:
                data = response.read()
                return json.loads(data.decode('utf-8'))

        except HTTPError as e:
            print(f"HTTP Error {e.code}: {e.reason}")
            print(f"URL: {url}")
            sys.exit(1)
        except URLError as e:
            print(f"URL Error: {e.reason}")
            print(f"URL: {url}")
            sys.exit(1)
        except Exception as e:
            print(f"Error fetching data: {e}")
            sys.exit(1)

    def format_date_danish(self, iso_date_string):
        """Convert ISO date string to Danish format (DD-MM-YYYY)

        Example: '2025-05-06T23:59:57.525504+02:00' -> '06-05-2025'
        """
        if not iso_date_string:
            return None

        try:
            # Parse ISO format (handles timezone)
            dt = datetime.fromisoformat(iso_date_string)
            return dt.strftime('%d-%m-%Y')
        except (ValueError, AttributeError):
            return None

    def extract_org_id_from_path(self, path):
        """Extract the last organization ID from the path

        Example: '1/2/6/55/131/' -> '131'
        """
        if not path:
            return None
        # Remove trailing slash and split
        parts = path.rstrip('/').split('/')
        # Return the last part
        return parts[-1] if parts else None

    def apply_org_name_override(self, org_name):
        """Apply organization name overrides if configured

        Example: 'Havet' -> 'Havet (underlejr)'
        """
        if not org_name:
            return org_name

        # Check for exact match override
        if org_name in self.ORG_NAME_OVERRIDES:
            return self.ORG_NAME_OVERRIDES[org_name]

        return org_name

    def clean_org_name(self, org_name):
        """Remove numeric prefix from organization name and apply overrides

        Example: '5532 - GRAS' -> 'GRAS'
        Returns: (numeric_id, clean_name)
        """
        if not org_name:
            return None, None

        # Match pattern: "5532 - GRAS" or just "GRAS"
        match = re.match(r'^(\d+)\s*-\s*(.+)$', org_name)
        if match:
            clean_name = match.group(2).strip()
            # Apply overrides after cleaning
            clean_name = self.apply_org_name_override(clean_name)
            return match.group(1), clean_name

        # No numeric prefix, just apply overrides
        clean_name = org_name.strip()
        clean_name = self.apply_org_name_override(clean_name)
        return None, clean_name

    def build_org_lookup(self, org_data):
        """Build organization name -> path lookup from org export data

        The org export has structure: [{"name": "5532 - GRAS", "parent_path": "1/2/6/55/131/"}, ...]
        We create a lookup by organization name to get the parent_path
        """
        print(f"Building organization lookup from {len(org_data)} organizations...")

        for org in org_data:
            org_name = org.get('name')
            org_path = org.get('parent_path')

            if org_name:
                # Store both by full name for exact matching
                self.org_lookup[org_name] = org_path

        print(f"Built lookup for {len(self.org_lookup)} organizations")

    def get_org_hierarchy_for_job(self, org_name):
        """Get full organizational hierarchy for a job

        Path structure:
        - Camp level: Ignored(1)/Lejr(2)/
        - Area level: Ignored(1)/Lejr(2)/Area(6)/
        - Committee level: Ignored(1)/Lejr(2)/Area(6)/Committee(55)/
        - Team level: Ignored(1)/Lejr(2)/Area(6)/Committee(55)/Team(131)/
        - Workgroup level: Ignored(1)/Lejr(2)/Area(6)/Committee(55)/Team(131)/Workgroup(456)/

        Returns: dict with area, committee, team, workgroup info
        """
        # Look up the org_path by organization name
        org_path = self.org_lookup.get(org_name)

        if not org_path:
            return {
                'area': 'Unknown',
                'area_full': 'Unknown',
                'committee': None,
                'committee_full': None,
                'team': None,
                'team_full': None,
                'workgroup': None,
                'workgroup_full': None,
                'full_path': [org_name]
            }

        # Parse the path: 1/2/6/55/131/456/
        parts = org_path.rstrip('/').split('/')

        # Remove first level (1=ignored)
        if len(parts) > 1:
            parts = parts[1:]  # Now: [2, 6, 55, 131, 456] or [2] for camp-level

        hierarchy = {
            'area': None,
            'area_full': None,
            'committee': None,
            'committee_full': None,
            'team': None,
            'team_full': None,
            'workgroup': None,
            'workgroup_full': None,
            'full_path': []
        }

        # Check if this is a camp-level job (only has ID 2)
        if len(parts) == 1 and parts[0] == '2':
            # Camp-level job - use special name
            clean_camp_name = self.clean_org_name(org_name)[1] or 'Spejdernes Lejr 2026'
            hierarchy['area'] = clean_camp_name
            hierarchy['area_full'] = org_name
            hierarchy['full_path'].append(clean_camp_name)
            return hierarchy

        # Remove camp level (2=lejr root) for regular jobs
        if len(parts) > 1:
            parts = parts[1:]  # Now: [6, 55, 131, 456]

        # We need to look up the org names for each ID in the path
        # Build a reverse lookup: org_id -> org_name
        org_id_lookup = {}
        for name, path in self.org_lookup.items():
            org_id = self.extract_org_id_from_path(path)
            if org_id:
                org_id_lookup[org_id] = name

        # Map path indices to levels
        # parts[0] = Area (level 3)
        # parts[1] = Committee (level 4) if exists
        # parts[2] = Team (level 5) if exists
        # parts[3] = Workgroup (level 6) if exists

        for idx, org_id in enumerate(parts):
            org_full_name = org_id_lookup.get(org_id, f"ID-{org_id}")
            org_numeric_id, org_clean_name = self.clean_org_name(org_full_name)

            if idx == 0:
                hierarchy['area'] = org_clean_name
                hierarchy['area_full'] = org_full_name
                hierarchy['full_path'].append(org_clean_name)
            elif idx == 1:
                hierarchy['committee'] = org_clean_name
                hierarchy['committee_full'] = org_full_name
                hierarchy['full_path'].append(org_clean_name)
            elif idx == 2:
                hierarchy['team'] = org_clean_name
                hierarchy['team_full'] = org_full_name
                hierarchy['full_path'].append(org_clean_name)
            elif idx == 3:
                hierarchy['workgroup'] = org_clean_name
                hierarchy['workgroup_full'] = org_full_name
                hierarchy['full_path'].append(org_clean_name)

        # Fallback to org name if nothing found
        if not hierarchy['area']:
            hierarchy['area'] = self.clean_org_name(org_name)[1] or 'Unknown'
            hierarchy['area_full'] = org_name

        return hierarchy

    def process_jobs(self, job_data):
        """Process jobs from API data"""
        print(f"Processing {len(job_data)} jobs...")

        jobs = []

        for job in job_data:
            job_id = job.get('id')
            job_name = job.get('name', 'Unnamed Job')
            org_name = job.get('organization_id', 'Unknown')

            # Get organizational hierarchy
            org_hierarchy = self.get_org_hierarchy_for_job(org_name)

            # Format dates
            create_date_iso = None
            create_date_danish = None
            if job.get('create_date'):
                create_date_iso = job.get('create_date')
                create_date_danish = self.format_date_danish(create_date_iso)

            # Build job object
            job_obj = {
                "id": job_id,
                "name": job_name,
                "teaser": job.get('teaser', ''),
                "description": job.get('description', ''),
                "description_time_and_scope": job.get('description_time_and_scope', ''),
                "requirements": job.get('requirements'),
                "application_count": job.get('application_count', 0),
                "no_of_recruitment": job.get('no_of_recruitment', 0),
                "no_of_hired_employee": job.get('application_count', 0),  # Using application_count as proxy
                "min_age": job.get('min_age'),
                "website_url": job.get('website_url', ''),
                "create_date": create_date_iso,
                "formatted_create_date": create_date_danish,
                "org_hierarchy": {
                    "area": org_hierarchy['area'],
                    "area_full": org_hierarchy['area_full'],
                    "committee": org_hierarchy['committee'],
                    "committee_full": org_hierarchy['committee_full'],
                    "team": org_hierarchy['team'],
                    "team_full": org_hierarchy['team_full'],
                    "workgroup": org_hierarchy['workgroup'],
                    "workgroup_full": org_hierarchy['workgroup_full']
                }
            }

            jobs.append(job_obj)

        print(f"Processed {len(jobs)} jobs")
        return jobs

    def write_output(self, jobs):
        """Write jobs data to JSON file"""
        output_path = Path(self.output_file)
        output_path.parent.mkdir(parents=True, exist_ok=True)

        with open(output_path, 'w', encoding='utf-8') as f:
            json.dump(jobs, f, ensure_ascii=False, indent=2)

        print(f"Wrote {len(jobs)} jobs to {self.output_file}")

    def run(self):
        """Main execution flow"""
        print("=" * 60)
        print("CampOS Job Export Fetcher")
        print("=" * 60)

        # Load configuration
        self.load_config()

        # Fetch organization data first (needed for hierarchy lookup)
        print("\nFetching organization data...")
        org_data = self.fetch_json(self.config['org_export_url'])
        print(f"Fetched {len(org_data)} organizations")

        # Build organization lookup
        self.build_org_lookup(org_data)

        # Fetch job data
        print("\nFetching job data...")
        job_data = self.fetch_json(self.config['job_export_url'])
        print(f"Fetched {len(job_data)} jobs")

        # Process jobs
        jobs = self.process_jobs(job_data)

        # Write output
        self.write_output(jobs)

        print("\n" + "=" * 60)
        print("Export complete!")
        print("=" * 60)


def main():
    """Main entry point"""
    # Get the script directory
    script_dir = Path(__file__).parent
    project_root = script_dir.parent

    # File paths
    config_file = script_dir / 'config.json'

    # Check for OUTPUT_DIR environment variable (for Docker)
    output_dir = os.getenv('OUTPUT_DIR')
    if output_dir:
        output_file = Path(output_dir) / 'jobs-export.json'
    else:
        # Default to project public directory
        output_file = project_root / 'public' / 'jobs-export.json'

    # Create processor and run
    processor = JobExportFetcher(config_file, output_file)
    processor.run()


if __name__ == '__main__':
    main()

File: scripts/fetch-job-export.js
#!/usr/bin/env node
/**
 * Fetch job export from CampOS API and generate jobs-export.json
 *
 * This script fetches data from the CampOS API:
 * - Job export: Contains job data with organization references
 * - Organization export: Contains organization hierarchy (parent_path)
 *
 * And generates:
 * - jobs-export.json (for use by the jobbank Vue app)
 * - Per-job OG HTML pages in public/job/{slug}/index.html
 *
 * Configuration:
 * - API credentials are stored in config.json (not committed to git)
 * - See config.example.json for the required structure
 *
 * Usage:
 *   node scripts/fetch-job-export.js
 *   npm run fetch-export
 */

import { readFileSync, writeFileSync, mkdirSync, rmSync, readdirSync, statSync, existsSync } from 'node:fs'
import { join, dirname } from 'node:path'
import { fileURLToPath } from 'node:url'
import { marked } from 'marked'
import { ORG_NAME_OVERRIDES } from '../src/config/org-overrides.js'

const __filename = fileURLToPath(import.meta.url)
const __dirname = dirname(__filename)

// ─── Config ──────────────────────────────────────────────────────────────────

function loadConfig(configFile) {
  let raw
  try {
    raw = readFileSync(configFile, 'utf8')
  } catch {
    console.error(`Error: Configuration file not found: ${configFile}`)
    console.error('Please create config.json based on config.example.json')
    process.exit(1)
  }

  let config
  try {
    config = JSON.parse(raw)
  } catch (e) {
    console.error(`Error: Invalid JSON in ${configFile}: ${e.message}`)
    process.exit(1)
  }

  const required = ['job_export_url', 'org_export_url']
  const missing = required.filter(f => !(f in config))
  if (missing.length) {
    console.error(`Error: Missing required fields in ${configFile}: ${missing.join(', ')}`)
    console.error('Please see config.example.json for the required structure')
    process.exit(1)
  }

  console.log(`Loaded configuration from ${configFile}`)
  return config
}

// ─── HTTP ─────────────────────────────────────────────────────────────────────

async function fetchJson(url) {
  let response
  try {
    response = await fetch(url, {
      headers: { 'User-Agent': 'JobBank/1.0' },
      signal: AbortSignal.timeout(30_000),
    })
  } catch (e) {
    console.error(`URL Error: ${e.message}`)
    console.error(`URL: ${url}`)
    process.exit(1)
  }

  if (!response.ok) {
    console.error(`HTTP Error ${response.status}: ${response.statusText}`)
    console.error(`URL: ${url}`)
    process.exit(1)
  }

  return response.json()
}

// ─── Date ─────────────────────────────────────────────────────────────────────

function formatDateDanish(isoDateString) {
  if (!isoDateString) return null
  // Extract date directly from the ISO string to preserve the source timezone
  const m = isoDateString.match(/^(\d{4})-(\d{2})-(\d{2})/)
  if (!m) return null
  return `${m[3]}-${m[2]}-${m[1]}`
}

// ─── Org hierarchy ────────────────────────────────────────────────────────────

function extractOrgIdFromPath(path) {
  if (!path) return null
  const parts = path.replace(/\/$/, '').split('/')
  return parts.length ? parts[parts.length - 1] : null
}

function applyOrgNameOverride(orgName) {
  if (!orgName) return orgName
  return ORG_NAME_OVERRIDES[orgName] ?? orgName
}

/** Returns [numericPrefix, cleanName] */
function cleanOrgName(orgName) {
  if (!orgName) return [null, null]
  // Match "5532 - GRAS", "0050 Korpsansatte", or just "GRAS"
  const m = orgName.match(/^(\d+)\s*(?:-\s*)?(.+)$/)
  if (m) {
    return [m[1], applyOrgNameOverride(m[2].trim())]
  }
  return [null, applyOrgNameOverride(orgName.trim())]
}

function buildOrgLookup(orgData) {
  console.log(`Building organization lookup from ${orgData.length} organizations...`)
  const orgLookup = {}
  for (const org of orgData) {
    if (org.name) orgLookup[org.name] = org.parent_path
  }
  console.log(`Built lookup for ${Object.keys(orgLookup).length} organizations`)
  return orgLookup
}

/**
 * Build a mapping from every org ID to its top-level area (short) name.
 *
 * Allows ?organization=nnn redirects to filter to the correct area.
 * Path structure:
 *   "1/"           -> root (skip)
 *   "1/2/"         -> camp level (skip)
 *   "1/2/9/"       -> area (ID 9 maps to itself)
 *   "1/2/9/27/"    -> committee (ID 27 maps to area 9)
 *   "1/2/9/27/28/" -> team (ID 28 maps to area 9)
 */
function buildOrgMap(orgLookup) {
  // Reverse lookup: org_id -> full_name
  const orgIdToName = {}
  for (const [name, path] of Object.entries(orgLookup)) {
    const orgId = extractOrgIdFromPath(path)
    if (orgId) orgIdToName[orgId] = name
  }

  const orgMap = {}
  for (const [name, path] of Object.entries(orgLookup)) {
    if (!path) continue
    const ownId = extractOrgIdFromPath(path)
    if (!ownId) continue

    // Strip the two fixed top levels: 1 (Spejderne root) and 2 (SL2026)
    let parts = path.replace(/\/$/, '').split('/')
    if (parts.length < 3) continue
    parts = parts.slice(2) // [area_id, ...]

    const areaFullName = orgIdToName[parts[0]]
    if (!areaFullName) continue

    const [, areaShortName] = cleanOrgName(areaFullName)
    if (areaShortName) orgMap[ownId] = areaShortName
  }

  console.log(`Built org_map for ${Object.keys(orgMap).length} org IDs`)
  return orgMap
}

function getOrgHierarchyForJob(orgName, orgLookup) {
  const orgPath = orgLookup[orgName]

  if (!orgPath) {
    return {
      area: 'Unknown', area_full: 'Unknown',
      committee: null, committee_full: null,
      team: null, team_full: null,
      workgroup: null, workgroup_full: null,
      full_path: [orgName],
    }
  }

  // Parse path "1/2/6/55/131/456/" and strip leading levels
  let parts = orgPath.replace(/\/$/, '').split('/')
  if (parts.length > 1) parts = parts.slice(1) // remove root (1)

  const hierarchy = {
    area: null, area_full: null,
    committee: null, committee_full: null,
    team: null, team_full: null,
    workgroup: null, workgroup_full: null,
    full_path: [],
  }

  // Camp-level job: only has ID 2
  if (parts.length === 1 && parts[0] === '2') {
    const cleanCampName = cleanOrgName(orgName)[1] || 'Spejdernes Lejr 2026'
    hierarchy.area = cleanCampName
    hierarchy.area_full = orgName
    hierarchy.full_path.push(cleanCampName)
    return hierarchy
  }

  if (parts.length > 1) parts = parts.slice(1) // remove camp level (2)

  // Reverse lookup: org_id -> org_name
  const orgIdLookup = {}
  for (const [name, path] of Object.entries(orgLookup)) {
    const orgId = extractOrgIdFromPath(path)
    if (orgId) orgIdLookup[orgId] = name
  }

  const levels = ['area', 'committee', 'team', 'workgroup']
  for (let i = 0; i < parts.length; i++) {
    const orgFullName = orgIdLookup[parts[i]] || `ID-${parts[i]}`
    const [, orgCleanName] = cleanOrgName(orgFullName)
    const level = levels[i]
    if (level) {
      hierarchy[level] = orgCleanName
      hierarchy[`${level}_full`] = orgFullName
      hierarchy.full_path.push(orgCleanName)
    }
  }

  if (!hierarchy.area) {
    hierarchy.area = cleanOrgName(orgName)[1] || 'Unknown'
    hierarchy.area_full = orgName
  }

  return hierarchy
}

// ─── Jobs ─────────────────────────────────────────────────────────────────────

function processJobs(jobData, orgLookup) {
  console.log(`Processing ${jobData.length} jobs...`)
  const jobs = []

  for (const job of jobData) {
    const orgName = job.organization_id || 'Unknown'
    const h = getOrgHierarchyForJob(orgName, orgLookup)
    const createDateIso = job.create_date || null

    jobs.push({
      id: job.id,
      name: job.name || 'Unnamed Job',
      teaser: job.teaser || '',
      description: job.description || '',
      description_time_and_scope: job.description_time_and_scope || '',
      requirements: job.requirements ?? null,
      application_count: job.application_count || 0,
      no_of_recruitment: job.no_of_recruitment || 0,
      no_of_hired_employee: job.application_count || 0,
      min_age: job.min_age ?? null,
      website_url: job.website_url || '',
      create_date: createDateIso,
      formatted_create_date: formatDateDanish(createDateIso),
      org_hierarchy: {
        area: h.area,
        area_full: h.area_full,
        committee: h.committee,
        committee_full: h.committee_full,
        team: h.team,
        team_full: h.team_full,
        workgroup: h.workgroup,
        workgroup_full: h.workgroup_full,
      },
    })
  }

  console.log(`Processed ${jobs.length} jobs`)
  return jobs
}

function writeOutput(jobs, orgMap, outputFile) {
  mkdirSync(dirname(outputFile), { recursive: true })
  writeFileSync(outputFile, JSON.stringify({ jobs, org_map: orgMap }, null, 2), 'utf8')
  console.log(`Wrote ${jobs.length} jobs and ${Object.keys(orgMap).length} org mappings to ${outputFile}`)
}

// ─── OG pages ─────────────────────────────────────────────────────────────────

function buildJobSlug(job) {
  let name = (job.name || '').toLowerCase()
  name = name.replace(/æ/g, 'ae').replace(/ø/g, 'oe').replace(/å/g, 'aa').replace(/é/g, 'e')
  name = name.replace(/[\u2011\u2013\u2014]/g, '-')
  name = name.replace(/[^a-z0-9-]+/g, '-')
  name = name.replace(/-{2,}/g, '-')
  name = name.replace(/^-|-$/g, '')
  return name ? `${name}-${job.id}` : String(job.id)
}

function escapeHtml(str) {
  return (str || '')
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
}

function stripHtmlTags(text) {
  return (text || '').replace(/<[^>]+>/g, '').trim()
}

/**
 * Generate per-job OG HTML pages alongside index.html.
 *
 * Reads the deployed index.html from outputDir, injects per-job OG meta tags,
 * and writes {outputDir}/job/{slug}/index.html for each job.
 *
 * Skips silently if index.html is not present (e.g. local dev where outputDir
 * is public/ and no build has run).
 */
function generateOgPages(jobs, outputDir) {
  const indexPath = join(outputDir, 'index.html')
  if (!existsSync(indexPath)) {
    console.log('Skipping OG page generation: index.html not found in output directory')
    return
  }

  const baseHtml = readFileSync(indexPath, 'utf8')
  const titleMatch = baseHtml.match(/<title>([^<]*)<\/title>/)
  const siteTitle = titleMatch ? titleMatch[1] : 'Jobbank - Spejdernes Lejr 2026'

  const siteUrl = (process.env.SITE_URL || 'https://jobs.spejderneslejr.dk').replace(/\/$/, '')
  const ogImageUrl = `${siteUrl}/og-jobbank.jpg`

  const jobDir = join(outputDir, 'job')
  mkdirSync(jobDir, { recursive: true })

  const generatedSlugs = new Set()

  for (const job of jobs) {
    const slug = buildJobSlug(job)
    generatedSlugs.add(slug)

    const jobName = escapeHtml(job.name || '')
    // Parse markdown → HTML → strip tags for plain-text OG description
    const teaserHtml = marked.parse(job.teaser || '')
    const teaser = escapeHtml(stripHtmlTags(teaserHtml).slice(0, 300))
    const canonicalUrl = `${siteUrl}/job/${slug}`

    let pageHtml = baseHtml.replace(
      /<title>[^<]*<\/title>/,
      `<title>${jobName} — ${escapeHtml(siteTitle)}</title>`
    )

    const ogMeta = [
      `  <meta property="og:type" content="website" />`,
      `  <meta property="og:site_name" content="${escapeHtml(siteTitle)}" />`,
      `  <meta property="og:title" content="${jobName}" />`,
      `  <meta property="og:description" content="${teaser}" />`,
      `  <meta property="og:url" content="${canonicalUrl}" />`,
      `  <meta property="og:image" content="${ogImageUrl}" />`,
      `  <meta name="twitter:card" content="summary_large_image" />`,
      `  <meta name="twitter:title" content="${jobName}" />`,
      `  <meta name="twitter:description" content="${teaser}" />`,
      `  <meta name="description" content="${teaser}" />`,
    ].join('\n') + '\n'

    pageHtml = pageHtml.replace('</head>', ogMeta + '</head>')

    const slugDir = join(jobDir, slug)
    mkdirSync(slugDir, { recursive: true })
    writeFileSync(join(slugDir, 'index.html'), pageHtml, 'utf8')
  }

  console.log(`Generated ${generatedSlugs.size} OG pages in ${jobDir}`)

  // Remove stale OG pages for jobs no longer in the export
  let staleCount = 0
  for (const entry of readdirSync(jobDir)) {
    const entryPath = join(jobDir, entry)
    if (statSync(entryPath).isDirectory() && !generatedSlugs.has(entry)) {
      rmSync(entryPath, { recursive: true })
      staleCount++
    }
  }
  if (staleCount) console.log(`Cleaned up ${staleCount} stale OG page(s)`)
}

// ─── Main ─────────────────────────────────────────────────────────────────────

async function main() {
  console.log('='.repeat(60))
  console.log('CampOS Job Export Fetcher')
  console.log('='.repeat(60))

  const scriptDir = __dirname
  const projectRoot = dirname(scriptDir)

  const configFile = join(scriptDir, 'config.json')
  const outputDir = process.env.OUTPUT_DIR
  const outputFile = outputDir
    ? join(outputDir, 'jobs-export.json')
    : join(projectRoot, 'public', 'jobs-export.json')

  const config = loadConfig(configFile)

  console.log('\nFetching organization data...')
  const orgData = await fetchJson(config.org_export_url)
  console.log(`Fetched ${orgData.length} organizations`)

  const orgLookup = buildOrgLookup(orgData)
  const orgMap = buildOrgMap(orgLookup)

  console.log('\nFetching job data...')
  const jobData = await fetchJson(config.job_export_url)
  console.log(`Fetched ${jobData.length} jobs`)

  const jobs = processJobs(jobData, orgLookup)
  writeOutput(jobs, orgMap, outputFile)
  generateOgPages(jobs, dirname(outputFile))

  console.log('\n' + '='.repeat(60))
  console.log('Export complete!')
  console.log('='.repeat(60))
}

main().catch(err => {
  console.error('Fatal error:', err)
  process.exit(1)
})
